projects/robots/festo/robotino3/controllers/robotino3/robotino3.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright 1996-2019 Cyberbotics Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Description: Simple avoidance controller and possibility to use the keyboard
+ *              to control manually the robot. A demo mode is also present.
+ *              The velocity of each wheel is set according to a
+ *              Braitenberg-like algorithm which takes the values returned
+ *              by the 9 infrared sensors as input.
+ */
+
+#include <webots/distance_sensor.h>
+#include <webots/keyboard.h>
+#include <webots/motor.h>
+#include <webots/robot.h>
+
+#include "base.h"
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define NUMBER_OF_INFRARED_SENSORS 9
+#define TIME_STEP (int)wb_robot_get_basic_time_step()  // from world file
+
+static int old_key = -1;
+static bool demo = false;
+static bool autopilot = true;
+static bool old_autopilot = true;
+static bool display_message = false;
+
+extern WbDeviceTag wheels[3];
+static WbDeviceTag infrared_sensors[NUMBER_OF_INFRARED_SENSORS];
+static const char *infrared_sensors_names[NUMBER_OF_INFRARED_SENSORS] = {"ir1", "ir2", "ir3", "ir4", "ir5",
+                                                                         "ir6", "ir7", "ir8", "ir9"};
+
+double convert_volt_to_meter(double V) {
+  // Don't forget to adapt the conversion according to the sensor model used
+  return 0.1594 * pow(V, -0.8533) - 0.02916;  // GP2D120
+  // return 0.7611 * pow(V, -0.9313) - 0.1252;   // GP2Y0A02YK0F
+  // return 0.1594 * pow(V, -0.8533) - 0.02916;  // GP2Y0A41SK0F
+  // return 20.24 * pow(V, -4.76) + 0.6632;      // GP2Y0A710K0F
```

Comment:
To make this a bit more generic and avoid commented lines, you should probably use the `wb_device_get_model` to apply the correct formula depending on the model of the sensors.